File: 01-programming-languages/C#/01-fundamentals/06-strings/strings.cs
// C# Strings

using System;

namespace StringsDemo
{
    class Program
    {
        static void Main(string[] args)
        {
            // String creation
            string s1 = "Hello";
            string s2 = new string('A', 5); // "AAAAA"
            string s3 = string.Concat("Hello", " ", "World");
            
            // String properties
            Console.WriteLine($"Length: {s1.Length}");
            
            // String methods
            Console.WriteLine($"Upper: {s1.ToUpper()}");
            Console.WriteLine($"Lower: {s1.ToLower()}");
            Console.WriteLine($"Trim: {"  hello  ".Trim()}");
            Console.WriteLine($"Substring: {s1.Substring(1, 3)}");
            Console.WriteLine($"Contains: {s1.Contains("ell")}");
            Console.WriteLine($"Starts with: {s1.StartsWith("He")}");
            Console.WriteLine($"Ends with: {s1.EndsWith("lo")}");
            
            // String formatting
            string name = "Alice";
            int age = 30;
            
            // String interpolation
            Console.WriteLine($"Name: {name}, Age: {age}");
            
            // Composite formatting
            Console.WriteLine("Name: {0}, Age: {1}", name, age);
            
            // String.Format
            string formatted = string.Format("Name: {0}, Age: {1}", name, age);
            
            // String concatenation
            string greeting = "Hello, " + name + "!";
            
            // StringBuilder (efficient for many operations)
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append("Hello");
            sb.Append(" ");
            sb.Append("World");
            string result = sb.ToString();
            
            // String split
            string csv = "apple,banana,cherry";
            string[] fruits = csv.Split(',');
            Console.WriteLine($"Fruits: {string.Join(", ", fruits)}");
            
            // String join
            string joined = string.Join(" - ", fruits);
            Console.WriteLine($"Joined: {joined}");
            
            // String replace
            string text = "Hello World";
            string replaced = text.Replace("World", "C#");
            Console.WriteLine($"Replaced: {replaced}");
            
            // String comparison
            string a = "apple";
            string b = "banana";
            Console.WriteLine($"Compare: {string.Compare(a, b)}");
            Console.WriteLine($"Equals: {a.Equals(b)}");
            
            // Verbatim string
            string path = @"C:\Users\Alice\Documents";
            Console.WriteLine($"Path: {path}");
            
            // Raw string literal (C# 11+)
            // string json = """
            // {
            //     "name": "Alice"
            // }
            // """;
            
            // String interpolation in verbatim
            string filePath = $@"C:\Users\{name}\Documents";
            
            // String.Contains with StringComparison
            Console.WriteLine($"Contains: {"Hello World".Contains("hello", StringComparison.OrdinalIgnoreCase)}");
            
            // String.IsNullOrEmpty
            string? nullStr = null;
            Console.WriteLine($"Is null or empty: {string.IsNullOrEmpty(nullStr)}");
            
            // String.IsNullOrWhiteSpace
            string whitespace = "   ";
            Console.WriteLine($"Is null or whitespace: {string.IsNullOrWhiteSpace(whitespace)}");
            
            Console.WriteLine("Strings example running");
        }
    }
}

File: 01-programming-languages/Python/01-fundamentals/06-strings/strings.py
"""String methods, formatting, and manipulation."""

# ── String Basics ────────────────────────────────────────────────────
s1 = "Hello, World"
s2 = 'Single quotes work too'
s3 = """Triple quotes
span multiple
lines"""

# ── String Operations ────────────────────────────────────────────────
print("Hello" + " " + "World")   # Concatenation
print("Ha" * 3)                  # Repetition: "HaHaHa"
print(len("Hello"))              # Length: 5
print("l" in "Hello")            # Membership: True

# ── Common Methods ───────────────────────────────────────────────────
s = "  Hello, World!  "
print(s.strip())         # "Hello, World!" (remove whitespace)
print(s.lower())         # "  hello, world!  "
print(s.upper())         # "  HELLO, WORLD!  "
print(s.title())         # "  Hello, World!  "
print(s.startswith("  "))  # True
print(s.endswith("!  "))   # True
print(s.find("World"))     # 9 (index, -1 if not found)
print(s.replace("World", "Python"))  # "  Hello, Python!  "
print(s.split(", "))       # ['  Hello', 'World!  ']

# ── Splitting and Joining ────────────────────────────────────────────
csv_line = "Alice,30,Engineer"
parts = csv_line.split(",")        # ['Alice', '30', 'Engineer']
rejoined = " | ".join(parts)       # "Alice | 30 | Engineer"

# Multi-character separator
text = "one::two::three"
parts = text.split("::")           # ['one', 'two', 'three']

# Split from right
"one.two.three".rsplit(".", 1)    # ['one.two', 'three']

# ── String Formatting ────────────────────────────────────────────────
name, age, score = "Alice", 30, 95.5

# f-strings (Python 3.6+) — preferred
print(f"{name} is {age} years old")
print(f"{score:.2f}")              # 95.50 (format float)
print(f"{name:>10}")              # "     Alice" (right-align)
print(f"{'hello':^20}")           # "       hello       " (center)
print(f"{255:08b}")               # 11111111 (binary)
print(f"{1000000:,}")             # 1,000,000 (comma separator)

# .format() method
print("{} is {} years old".format(name, age))
print("{name} is {age} years old".format(name="Alice", age=30))

# % formatting (old style)
print("%s is %d years old" % (name, age))

# ── String Methods Continued ─────────────────────────────────────────
s = "Hello123"
print(s.isalnum())      # True (alphanumeric)
print(s.isalpha())      # False (contains digits)
print(s.isdigit())      # False (contains letters)
print(s.isnumeric())    # False
print(s.istitle())      # True
print(s.isupper())      # False

# ── Encoding ─────────────────────────────────────────────────────────
text = "Hello"
encoded = text.encode("utf-8")    # b'Hello' (bytes)
decoded = encoded.decode("utf-8") # "Hello" (str)

# ── Useful Patterns ──────────────────────────────────────────────────
# Check if string is palindrome
def is_palindrome(s):
    clean = s.lower().replace(" ", "")
    return clean == clean[::-1]

# Reverse a string
reversed_str = "hello"[::-1]  # "olleh"

# Count words
word_count = len("hello world foo".split())

# Title case without small words
def smart_title(s):
    small_words = {"a", "an", "the", "in", "on", "at", "for", "of"}
    words = s.split()
    return " ".join(
        w if w.lower() in small_words and i > 0 else w.title()
        for i, w in enumerate(words)
    )

# Remove all whitespace
no_spaces = "".join("hello world".split())

# Truncate with ellipsis
def truncate(s, max_len):
    return s[:max_len-1] + "..." if len(s) > max_len else s
